app/uk/gov/hmrc/cgtpropertydisposals/service/TaxEnrolmentService.scala

```diff
@@ -18,60 +18,72 @@ package uk.gov.hmrc.cgtpropertydisposals.service
+//import cats.implicits._
 import com.google.inject.{ImplementedBy, Inject, Singleton}
-import play.api.http.Status._
 import uk.gov.hmrc.cgtpropertydisposals.connectors.TaxEnrolmentConnector
-import uk.gov.hmrc.cgtpropertydisposals.models.{Address, EnrolmentRequest, Error, KeyValuePair, SubscriptionDetails}
-import uk.gov.hmrc.cgtpropertydisposals.service.TaxEnrolmentService.TaxEnrolmentResponse
-import uk.gov.hmrc.cgtpropertydisposals.service.TaxEnrolmentService.TaxEnrolmentResponse.TaxEnrolmentCreated
-import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.cgtpropertydisposals.models.{Error, TaxEnrolmentRequest}
+import uk.gov.hmrc.cgtpropertydisposals.repositories.TaxEnrolmentRepository
+import uk.gov.hmrc.cgtpropertydisposals.util.Logging
+import uk.gov.hmrc.http.{HeaderCarrier, HttpResponse}
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 @ImplementedBy(classOf[TaxEnrolmentServiceImpl])
 trait TaxEnrolmentService {
-  def allocateEnrolmentToGroup(cgtReference: String, subscriptionDetails: SubscriptionDetails)(
+  def allocateEnrolmentToGroup(taxEnrolmentRequest: TaxEnrolmentRequest)(
     implicit hc: HeaderCarrier
-  ): EitherT[Future, Error, TaxEnrolmentResponse]
-}
-
-object TaxEnrolmentService {
-
-  sealed trait TaxEnrolmentResponse
-
-  object TaxEnrolmentResponse {
-    case object TaxEnrolmentCreated extends TaxEnrolmentResponse
-  }
-
+  ): EitherT[Future, Error, Unit]
 }
 
 @Singleton
-class TaxEnrolmentServiceImpl @Inject()(connector: TaxEnrolmentConnector) extends TaxEnrolmentService {
+class TaxEnrolmentServiceImpl @Inject()(
+  taxEnrolmentConnector: TaxEnrolmentConnector,
+  taxEnrolmentRetryRepository: TaxEnrolmentRepository
+) extends TaxEnrolmentService
+    with Logging {
 
-  override def allocateEnrolmentToGroup(cgtReference: String, subscriptionDetails: SubscriptionDetails)(
+  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  override def allocateEnrolmentToGroup(taxEnrolmentRequest: TaxEnrolmentRequest)(
     implicit hc: HeaderCarrier
-  ): EitherT[Future, Error, TaxEnrolmentResponse] = {
+  ): EitherT[Future, Error, Unit] = {
 
-    val enrolmentRequest = subscriptionDetails.address match {
-      case Address.UkAddress(line1, line2, line3, line4, postcode) =>
-        EnrolmentRequest(List(KeyValuePair("Postcode", postcode)), List(KeyValuePair("CGTPDRef", cgtReference)))
-      case Address.NonUkAddress(line1, line2, line3, line4, maybePostcode, countryCode) =>
-        EnrolmentRequest(List(KeyValuePair("CountryCode", countryCode.code)), List(KeyValuePair("CGTPDRef", cgtReference)))
-    }
+    val taxEnrolmentResponse: EitherT[Future, Error, Unit] = for {
+      httpResponse <- taxEnrolmentConnector
+                       .allocateEnrolmentToGroup(taxEnrolmentRequest)
+                       .leftMap(e => Error(s"Error calling tax enrolment service: $e"))
+      result <- EitherT.fromEither(handleTaxEnrolmentResponse(httpResponse))
```

Comment:
you can probably do a `.leftFlatMap` or `.leftSemiFlatMap` here to avoid having to to the `EitherT.leftF` below

Reply:
do you prefer this:

```scala
    val taxEnrolmentResponse: EitherT[Future, Error, Unit] = for {
      httpResponse <- taxEnrolmentConnector
                       .allocateEnrolmentToGroup(taxEnrolmentRequest)
                       .leftMap(e => Error(s"Error calling tax enrolment service: $e"))
      result <- EitherT.fromEither(handleTaxEnrolmentResponse(httpResponse))
    } yield result

    taxEnrolmentResponse.biSemiflatMap(
      (error: Error) => {
        logger.warn(s"Failed to allocate enrolments due to error: $error. Inserting enrolment details in mongo.")
        val dbResponse = for {
          writeResult <- taxEnrolmentRetryRepository
                          .insert(taxEnrolmentRequest)
                          .leftMap(error => Error(s"Error inserting enrolment details into mongo: $error"))
        } yield (writeResult)

        dbResponse.map { writeResult =>
          if (writeResult) Future.successful(error)
          else Future.successful(Error("Failed to insert enrolment details into mongo"))
        }
        Future.successful(error)
      },
      _ => Future.successful(())
    )
```

Reply:
I meant having the left mapping inside the for comprehension itself so that you didn't have to handle the right case since you don't want to touch that

Reply:
gotcha - misunderstood your initial comment